DUI: test for png files

aosyatnik opened the pull request:
As it was discussed, we need to create test, that can check png files.
This test takes all entries and checks icon file. If icon is not found test logs icon's title and fails.

Reviewers
@Benglin , please take a look.

Link to YouTrack:
[MAGN-6555](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6555) DUI: Create tests for checking icons png files

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,95 @@
+            };
+
+            var preloader = new Preloader(rootFolder, versions);
+            preloader.Preload();
```
The `preloader.Preload()` method will cause ASM binaries to be loaded, but that's not needed for these tests, so please remove that `Preload` call.

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,95 @@
+        static readonly string binPath = Directory.GetParent(AnyCPUPath).FullName;
+        static readonly string DynamoPath = Directory.GetParent(binPath).FullName;
+
+        static readonly string ResourcePath = Path.Combine(DynamoPath, @"src\Resources");
```
The above steps can be folded into a helper method `InitializeDirectoryPaths` so that we can assert that nothing changes in those assumptions. Perhaps it can do this:

```csharp
var assemblyDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
var relativeDir = Path.Combine(asmDir, @"..\..\..\src\Resources");
var absoluteDir = Path.GetFullPath(relativeDir);
Assert.IsFalse(string.IsNullOrEmpty(absoluteDir));
Assert.IsTrue(Directory.Exists(absoluteDir));
```

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,95 @@
+
+            foreach (var entry in searchEntries)
+            {
+                if (entry is NodeSearchElement)
```
```csharp
if (!(entry is NodeSearchElement))
    continue;
```

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,95 @@
+
+                    string largeIconPath = Path.Combine(ResourcePath, Path.GetFileNameWithoutExtension(searchEle.Assembly),
+                        "LargeIcons", searchEle.IconName + Configurations.LargeIconPostfix + ".png");
+
```
The above two have the same part:

`Path.Combine(ResourcePath, Path.GetFileNameWithoutExtension(searchEle.Assembly)`, maybe it can be stored as such:

```csharp
var prefix = Path.Combine(ResourcePath, Path.GetFileNameWithoutExtension(searchEle.Assembly);
string smallIconPath = Path.Combine(prefix, "SmallIcons", ...);
string largeIconPath = Path.Combine(prefix, "LargeIcons", ...);
```

Does that make sense?

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,95 @@
+                        "LargeIcons", searchEle.IconName + Configurations.LargeIconPostfix + ".png");
+
+                    bool smallIconExists = File.Exists(smallIconPath);
+                    bool largeIconExists = File.Exists(largeIconPath);
```
Instead of checking for physical files, I'm not sure if we can go through the `LibraryServices` to perform actual icon loading? That will validate the whole resource locating logic. How does that sound?

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,95 @@
+                    if (!largeIconExists)
+                        Debug.WriteLine(largeIconPath);
+
+                    testResult = testResult && smallIconExists && largeIconExists;
```
I think we can maintain a list of icons that are not found, say, in `List<string> missingIcons`, and then in the end assert that the list is empty, if it is not empty, then display the content of that `missingIcons` in one go:

```csharp
// If there is any missing icon, display them in one go in NUnit output.
string missingIconNames = TurnListIntoOneString(missingIcons);
Assert.IsTrue(missingIcons.Length == 0, missingIconNames);
```

Benglin commented:
Thanks a lot for doing this, @aosyatnik. I think this is a really good start! Let's see if you can make the changes I think are slightly better.

aosyatnik commented:
Hi, @Benglin , I used IconServices to check whether icon is presented or not. Please, take a look.

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,104 @@
+                {
+                    GeometryFactoryPath = geometryFactoryPath,
+                    Preferences = PreferenceSettings.Load()
+                });
```
Hi @aosyatnik, I think `PreferenceSettings.Load` is removed. If that's the case, just ignore `Preferences` and leave it as `null`, it will be created internally.

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,104 @@
+            };
+
+            var preloader = new Preloader(rootFolder, versions);
+            geometryFactoryPath = preloader.GeometryFactoryPath;
```
This can be further simplified as:

```csharp
var preloader = new Preloader(rootFolder);
geometryFactoryPath = preloader.GeometryFactoryPath;
```

And then remove `var versions`.

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,104 @@
+            foreach (var entry in searchEntries)
+            {
+                if (!(entry is NodeSearchElement))
+                    continue;
```
This can be removed if we do this above:

`searchEntries = model.SearchModel.SearchEntries.OfType<NodeSearchElement>();`

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,104 @@
+    {
+        private static string geometryFactoryPath = "";
+        private IEnumerable searchEntries;
+        private IconServices iconServices = new IconServices();
```
I think `IconServices` can only be instantiated with an `IPathManager` now, you can get it from `DynamoModel.PathManager` property after `DynamoModel` is created.

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,104 @@
+    internal class IconsTests
+    {
+        private static string geometryFactoryPath = "";
+        private IEnumerable searchEntries;
```
`searchEntries` and `geometryFactoryPath` are not needed here.

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        [SetUp]
+        public void PreloadShapeManager()
```
Please merge both `InitializeDirectoryPaths` and `PreloadShapeManager` into one method called `SetupIconTests` and have another one called `TearDownIconTests` (which is going to be empty I guess, or it can mark `DynamoModel` or `Preloader` instances as `null`).

Benglin commented on test/DynamoCoreTests/IconsTests.cs:
```diff
@@ -0,0 +1,104 @@
+                var largeIconName = searchEle.IconName + Configurations.LargeIconPostfix;
+
+
+                var warehouse = iconServices.GetForAssembly(searchEle.Assembly);
```
We can move these out of the `foreach` loop:

```csharp
var currentWarehouseAssembly = string.Empty;
IconWarehouse currentWarehouse = null;
```

Then in the loop:

```csharp
foreach(...)
{
    if (currentWarehouseAssembly != searchEle.Assembly) {
        currentWarehouseAssembly = searchEle.Assembly;
        currentWarehouse = iconServices.GetForAssembly(searchEle.Assembly);
    }
}
```

Does that make sense?

Benglin commented:
Again, @aosyatnik, this is really a good work, I can't wait for it to go green. Please update these and I'll merge it (currently it breaks the build). Thanks :)

aosyatnik commented:
@Benglin , I reworked this code. I found out, that for creation dynamo model we don't need factory path or anything else. So I removed all set up methods. Now test is clearer.

Benglin commented:
Thanks for the cleaner test, @aosyatnik! I think you didn't quite get the intention of [my previous comment](https://github.com/DynamoDS/Dynamo/pull/3938#discussion_r26549927). I would prefer you not to call `iconServices.GetForAssembly` that frequently, that's why I introduced `string currentWarehouseAssembly` :)

aosyatnik commented:
Hi, @Benglin , I was inattentive. I've added `string currentWarehouseAssembly`.

Benglin commented:
Thanks again, @aosyatnik! Sorry for being nitpicking! LGTM!
